app/services/category_hashtag_data_source.rb

```diff
@@ -52,11 +52,43 @@ def self.search(guardian, term, limit)
   end
 
   def self.search_sort(search_results, term)
-    search_results
-      .select { |item| item.slug == term }
-      .sort_by { |item| item.text.downcase }
-      .concat(
-        search_results.select { |item| item.slug != term }.sort_by { |item| item.text.downcase },
-      )
+    if term.present?
+      search_results
+        .select { |item| item.slug == term }
+        .sort_by { |item| item.text.downcase }
+        .concat(
+          search_results.select { |item| item.slug != term }.sort_by { |item| item.text.downcase },
+        )
+    else
+      search_results.sort_by { |item| item.text.downcase }
+    end
+  end
+
+  def self.search_without_term(guardian, limit)
+    guardian_categories = Site.new(guardian).categories
```

Comment:
`Site#categories` does alot of things which I'm not sure is necessary for the intended use case here. I think we can rely on `User#secure_category_ids` instead.

Reply:
The main benefit of `Site.new(guardian).categories` is that all the categories are already cached and we just scope them down to the user, so I don't think there are additional queries involved. I have used it in the other places in `CategoryHashtagDataSource` too. The other benefit is that it already figures out parent category details. But maybe it does make more sense to use that here and in the other places if the query time is negligible?

Reply:
I think the method is partially cached and it only memoizes the results per instance. Since this is a new instance, we're not going to be benefit from the memoization here. The method does execute extra queries even after fetching all the categories from the class level cache: 

https://github.com/discourse/discourse/blob/e669cf364893421b6545312f62ba7ad5299ee725/app/models/site.rb#L102-L140

Reply:
Hmm okay I will rethink this, it's probably fine to just use `User#secure_category_ids` then...thanks for pointing this out!

Reply:
I actually would prefer to refactor this in a separate PR too so it's encapsulated as a separate single commit.